utils/dataset_manifest/core.py

```diff
@@ -139,23 +140,33 @@ def __iter__(self) -> Iterator[Union[int, Tuple[int, int, str]]]:
 
 class DatasetImagesReader:
     def __init__(self,
-                sources: Union[List[str], List[BytesIO]],
-                *,
-                start: int = 0,
-                step: int = 1,
-                stop: Optional[int] = None,
-                meta: Optional[Dict[str, List[str]]] = None,
-                sorting_method: SortingMethod =SortingMethod.PREDEFINED,
-                use_image_hash: bool = False,
-                **kwargs):
-        self._raw_data_used = not isinstance(sources[0], str)
-        func = (lambda x: x.filename) if self._raw_data_used else None
-        self._sources = sort(sources, sorting_method, func=func)
+        sources: Union[List[str], Iterator[BytesIO]],
+        *,
+        start: int = 0,
+        step: int = 1,
+        stop: Optional[int] = None,
+        meta: Optional[Dict[str, List[str]]] = None,
+        sorting_method: SortingMethod =SortingMethod.PREDEFINED,
+        use_image_hash: bool = False,
+        are_sources_filtered: bool = False,
+        **kwargs
+    ):
+        self._is_generator_used = isgenerator(sources)
+
+        if not self._is_generator_used:
+            raw_data_used = not isinstance(sources[0], str)
+            func = (lambda x: x.filename) if raw_data_used else None
+            self._sources = sort(sources, sorting_method, func=func)
+        else:
+            if sorting_method != SortingMethod.PREDEFINED:
+                raise ValueError('Only SortingMethod.PREDEFINED can be used with generator')
```

Comment:
No, we cannot apply sorting in that case because we don't want to keep all files in memory to perform sorting. Besides that, `ThreadPoolExecutor.map` keeps sequence order and the list of files is sorted before downloading step.

Reply:
What do you mean by "keep all the files in the memory"? Getting the list of files from the storage should be enough for sorting and/or filtering the range. After the requested segment of the list is obtained, it can be sorted as requested. Then, the actual images can be downloaded for further processing.

Reply:
If the list of files is first obtained from the storage, gets filtered, sorted, and then the requested files are downloaded and passed here in a generator, it's ok.